Stop reporting changes for configured alias (#51068)

lvinagre opened the pull request:
### What does this PR do?
This fixes an issue where Salt is reporting changes for networking aliases when they are already configured as expected and no changes are done. 

### What issues does this PR fix or reference?
Fixes: #51068 

### Previous Behavior
```
Thu Nov 24 10:00:01 EST 2022
local:
----------
          ID: pub_eth_alias
    Function: network.managed
        Name: eno1:11
      Result: True
     Comment: Interface eno1:11 is up to date.
     Started: 10:01:47.468357
    Duration: 53.915 ms
     Changes:   
              ----------
              status:
                  Interface eno1:11 is up
```

### New Behavior
```
Thu Nov 24 15:25:32 EST 2022
local:
----------
          ID: pub_eth_alias
    Function: network.managed
        Name: eno1:11
      Result: True
     Comment: Interface eno1:11 is up to date.
     Started: 15:27:14.110393
    Duration: 48.243 ms
     Changes:   
```


### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

I am not good at writing tests, but at least the current ones are passing:
```
==================================================================================== test session starts ====================================================================================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-0.13.0                       
tempdir: /tmp/stsuite                                                                         
max open files; soft: 3072; hard: 1048576                                                     
rootdir: /home/vagrant/salt, configfile: pytest.ini                                                                                                                                          plugins: subtests-0.4.0, skip-markers-1.2.0, custom-exit-code-0.3.0, helpers-namespace-2021.4.29, shell-utilities-1.6.0, system-statistics-1.0.2, timeout-1.4.2, flaky-3.7.0, httpserver-1.0.
4, tempdir-2019.10.12, salt-factories-1.0.0rc21                                                                                                                                              
collected 3 items                                                                                                                                                                           
                                                                                              
tests/unit/states/test_network.py ...                                                         
                                                                                                                                                                                             
===================================================================================== warnings summary ======================================================================================
tests/conftest.py:229                                                                                                                                                                        
  /home/vagrant/salt/tests/conftest.py:229: PytestDeprecationWarning: The hookimpl pytest_configure uses old-style configuration options (marks or attributes).
  Please use the pytest.hookimpl(trylast=True) decorator instead
   to configure the hooks.                                                                                                                                                                   
   See https://docs.pytest.org/en/latest/deprecations.html#configuring-hook-specs-impls-using-markers
    @pytest.mark.trylast         
                                                                                              
tests/unit/states/test_network.py::NetworkTestCase::test_managed                                                                                                                               /home/vagrant/salt/.nox/test-parametrized-3-crypto-pycryptodome-transport-tcp-coverage-false/lib/python3.9/site-packages/napalm/base/helpers.py:8: DeprecationWarning: Using or importing t
he ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import Iterable       
                                                                                              
tests/unit/states/test_network.py::NetworkTestCase::test_managed                                                                                                                               /home/vagrant/salt/.nox/test-parametrized-3-crypto-pycryptodome-transport-tcp-coverage-false/lib/python3.9/site-packages/jnpr/junos/factcache.py:10: DeprecationWarning: Using or importing
 the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    class _FactCache(collections.MutableMapping):          
                                                                                              
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html                                                                                                                      
=============================================================================== 3 passed, 3 warnings in 1.01s ===============================================================================
nox > Session test-parametrized-3(crypto='pycryptodome', transport='tcp', coverage=False) was successful.                                                                                    
```

### Commits signed with GPG?
No

lvinagre commented:
Hey @waynew , just want to check if there's anything missing in that PR for review.

Ch3LL commented on changelog/51068.fixed:
```diff
@@ -0,0 +1 @@
+Made network.managed stop reporting changes when there are none.
```
The file extension needs to be .md now. So (51068.fixed.md)

Ch3LL commented on salt/states/network.py:
```diff
@@ -595,30 +595,16 @@ def managed(name, enabled=True, **kwargs):
+        if name.split(':')[0] in interfaces:
+            interface_status = interfaces[name.split(':')[0]].get('up')
         else:
-            for iface in interfaces:
```
Can you elaborate on why you are removing this section?

lvinagre replied:
Let's see if I still remember that :sweat_smile: 

So, although we get that nested result in the network.interfaces call, yet, when managing it with network.managed the name will always be either {interface}:{X} for the alias or any other simple name that does not match the main interface name, which AFAIR would be considered a different interface.

With that in mind, there is no real need to go through whatever is nested on an alias, as the main interface will either be up or down and the same would apply to the rest that is not nested. Otherwise (if it is not an alias to an interface and not an interface itself), it is probably not configured and hence down.

Ch3LL replied:
Sounds good. Looks like we just need the changelog and test coverage to get this over the line

lvinagre replied:
I've spent some time reading about the test part and tried writing some new and improving what was present. It runs fine for me, but probably worth a second look. Also, changelog extension updated (included .md).
```
==================================================================================== test session starts ====================================================================================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-0.13.0 -- /repo/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/bin/python
cachedir: .pytest_cache
tempdir: /tmp/stsuite
max open files; soft: 3072; hard: 1048576
rootdir: /repo, configfile: pytest.ini
plugins: subtests-0.4.0, skip-markers-1.2.0, custom-exit-code-0.3.0, helpers-namespace-2021.4.29, shell-utilities-1.6.0, system-statistics-1.0.2, timeout-1.4.2, flaky-3.7.0, httpserver-1.0.4, tempdir-2019.10.12, salt-factories-1.0.0rc21
collected 7 items                                                                                                                                                                           

tests/integration/states/test_network.py::NetworkTest::test_managed_addInterface0 PASSED
tests/integration/states/test_network.py::NetworkTest::test_managed_addInterface1 PASSED
tests/integration/states/test_network.py::NetworkTest::test_managed_existingInterfaceNoChanges0 PASSED
tests/integration/states/test_network.py::NetworkTest::test_managed_existingInterfaceWithChanges0 PASSED
tests/integration/states/test_network.py::NetworkTest::test_managed_existingInterfaceWithChanges1 PASSED
tests/integration/states/test_network.py::NetworkTest::test_routes PASSED
tests/integration/states/test_network.py::NetworkTest::test_system PASSED

===================================================================================== warnings summary ======================================================================================
tests/integration/states/test_network.py::NetworkTest::test_managed_addInterface0
  /repo/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/napalm/base/helpers.py:8: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import Iterable

tests/integration/states/test_network.py::NetworkTest::test_managed_addInterface0
  /repo/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/jnpr/junos/factcache.py:10: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    class _FactCache(collections.MutableMapping):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================== 7 passed, 2 warnings in 33.13s ===============================================================================
```
Let me know if there is anything else from me here and thanks for the help! :+1: :smile:

Ch3LL reviewed (changes requested):
This will also require test coverage

Ch3LL commented on tests/integration/states/test_network.py:
```diff
@@ -25,18 +30,130 @@ def setUp(self):
+
+        # and validate results
+        self.assertSaltTrueReturn(ret)
+        self.assertInSaltComment('Interface {} is up to date.'.format(iface_name), ret)
```
It looks like these tests rely on the previous tests running before them. Is that assumption correct? Is so we want to ensure that is not the case.

lvinagre replied:
I did not find a lot of documentation on how to setup the tests or use stuff (e.g. salt_master, salt_call_cli etc) with the new pytests but figured out most from pytest docs and the current files inside the `tests/pytests` folder. I moved the tests there and results are still good.

Related to the concern about the dependency of tests, they do not rely on previous. Instead, I put in place a common function to add the same stuff on each test when needed and used the fixture for the things that are common across all of them.

```
==================================================================================== test session starts ====================================================================================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-0.13.0 -- /repo/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/bin/python
cachedir: .pytest_cache
tempdir: /tmp/stsuite
max open files; soft: 3072; hard: 1048576
rootdir: /repo, configfile: pytest.ini
plugins: subtests-0.4.0, skip-markers-1.2.0, custom-exit-code-0.3.0, helpers-namespace-2021.4.29, shell-utilities-1.6.0, system-statistics-1.0.2, timeout-1.4.2, flaky-3.7.0, httpserver-1.0.4, tempdir-2019.10.12, salt-factories-1.0.0rc21
collected 10 items                                                                                                                                                                          

tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_addInterface0 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_addInterface1 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_existingInterfaceNoChanges0 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_existingInterfaceNoChanges1 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_existingInterfaceWithChanges0 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_existingInterfaceWithChanges1 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_existingInterfaceWithChanges2 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_existingInterfaceWithChanges3 PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_routes PASSED
tests/pytests/integration/states/test_network.py::TestNetwork::test_system PASSED

===================================================================================== warnings summary ======================================================================================
tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_addInterface0
  /repo/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/napalm/base/helpers.py:8: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import Iterable

tests/pytests/integration/states/test_network.py::TestNetwork::test_managed_addInterface0
  /repo/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/jnpr/junos/factcache.py:10: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    class _FactCache(collections.MutableMapping):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================== 10 passed, 2 warnings in 168.34s (0:02:48) =========================================================================
nox > Session test-parametrized-3(crypto=None, transport='zeromq', coverage=False) was successful.
nox > Ran multiple sessions:
nox > * test-3(coverage=False): success
nox > * test-parametrized-3(crypto=None, transport='zeromq', coverage=False): success
```

Ch3LL reviewed (changes requested):
We actually are migrating all of our tests to pytests in `tests/pytests`. Any chance you would be willing to add these new tests there instead?

Ch3LL commented:
@lvinagre looks like there are some failing tests related to your changes.

lvinagre commented:
> @lvinagre looks like there are some failing tests related to your changes.

@Ch3LL, I checked the previous tests and fixed some stuff, mainly:
- Fixed a bug on the teardown for rhel
- Added some handling on the values for the test for rhel (some of the values used on debian were no good there)
- Fixed a clause that was preventing tests to run properly debian like systems (e.g. Ubuntu)

Probably a good idea to run tests again before merging?

lvinagre commented:
Just had the time to review the checks output this time and I do not see any caused by the test themselves. Some have timed out in factory classes, there is a windows one that failed on the setup and all the others that failed on the tests did so because "ifup" is not present, but this is not a requirement from the tests or the state, but from the network module which is used within the state.

Ch3LL commented:
Are you able to replicate the test failures locally? These tests are not failing on the head of the branch so something must be related to these changes because the network tests are failing consistently. Also looks like pre-commit and lint are failing.

lvinagre commented:
Hi @Ch3LL. They do work fine here, naturally for Debian and with a push, installing the network-scripts, on CentOS. Though the latter is not a requirement from my tests, but from the network module itself.

I can see why they are not failing on the head branch. The current file in the head branch only has 3 tests which are all using the "test=true" flag and I am adding a couple of test which run for real in addition to those 3, hence the ones that carry the flag are working fine because the do not get to touch anything and the "real" ones are failing in CentOS based systems as they do not have the networks-scripts package installed by default (I did not think it was a good idea to add them in the tests as they are required to use the module/state).

I am updating the commit again to fix the lint issues, not sure I should touch anything to the pre-commit part.

Ch3LL commented:
Okay, I can try to dive into this myself but it may be a week or two before I have the bandwidth.

lvinagre commented:
Hey @Ch3LL , just checking back if you had the time check what you wanted here. Also, please let me know if I can be of further help. :+1:

Ch3LL commented:
Apologies, I have not been able to get this, but it is on my list of PRs to follow up on personally.

twangboy commented:
Looks like you have some failing tests.

lvinagre committed:
Stop reporting changes for configured alias (#51068)

lvinagre commented:
@twangboy, I spent a bit more time on this now and was able to find some stuff that was actually broken. I've put all into a single commit, rebased and put it through pre-commit and it looks good. I've also tested the final thing in in the distros that were failing (debian 11/12 and rocky 8/9) and all tests passed. Hopefully now we are good to go, my apologies for the previous/broken stuff.

twangboy commented:
Please resolve the conflicts

lvinagre commented:
> Please resolve the conflicts

done

dwoz committed:
Add unit test for alias no-spurious-change regression (#51068)

twangboy commented on salt/states/network.py:
```diff
@@ -621,30 +621,16 @@ def managed(name, enabled=True, **kwargs):
-                                    interface_status = True
+            interface_status = False
+
+        # Handle interface as desired
```
Unconditionally splitting the interface string on a colon breaks support for complex interface naming conventions (like VLAN tags and sub-interfaces) and incorrectly conflates a parent interface's status with its un-validated alias configurations.

By completely deleting that section and substituting the raw split(":"), the code now blindly conflates the parent interface's operational status with the status of the alias itself. If the parent interface (eth0) is up, Salt will now automatically assume the alias (eth0:1) is up and up-to-date—even if the virtual IP allocation for that alias completely failed to bind on the host.

That deleted code was the only mechanism verifying that the specific label (second.get("label") == name) was genuinely active in the system's runtime network state.

Instead of an unconditional split(":"), Salt should use a targeted regex match to separate a true legacy sub-interface/alias label from a standard interface name. Then, it needs to query the actual nested secondary data structure to confirm that the specific virtual label is bound and active.

```python
    # Bring up/shutdown interface
    try:
        # Get Interfaces information
        interfaces = salt.utils.network.interfaces()
        interface_status = False

        # 1. Check if the exact interface name exists directly
        if name in interfaces:
            interface_status = interfaces[name].get("up", False)
        
        # 2. Safely check for legacy alias structures (e.g., eth0:1, bond0:vlan1)
        else:
            import re
            # Match standard interface naming followed by a numeric or label alias
            match = re.match(r"^([a-zA-Z0-9.\-_]+):([a-zA-Z0-9_\-]+)$", name)
            if match:
                parent_iface = match.group(1)
                if parent_iface in interfaces:
                    # Look for the explicit alias label in the secondary IP data structures
                    for family in ("secondary", "inet", "inet6"):
                        if family in interfaces[parent_iface]:
                            for data in interfaces[parent_iface][family]:
                                if data.get("label") == name:
                                    interface_status = True
                                    break
                        if interface_status:
                            break

        interface_status = interface_status or False

        # Handle interface as desired
        if enabled:
```

twangboy reviewed (changes requested):
This should be created on the earliest supported branch where the bug exists. Probably 3006.x
